### PR TITLE
Status display of current job

### PR DESCRIPTION
**Pull request by drothlis**
It looks like this:

current-status-display

TODO:

- [x] Back-end reports start & end times from memory
- [x] Self-tests for start/end time
- [x] Back-end reports test pass/fail/error counts as they come in
- [x] Front-end displays it
- ~~Back-end caches mongo query of pass/fail/error counts~~ (leave it until we implement server-sent events)
- [x] Back-end remembers the above across restarts
- [x] Convert from UTC to user's timezone, in the browser (timezone reported by the browser?)
- ~~e2e-tests~~
- [x] [Encode start & end timestamps as ISO 8601](http://flask.pocoo.org/snippets/119/)
- Address comments from Will's review:
  - [x] I'd like to see some tests for the start time/end time stuff in service.
  - [x] I think it'd be better to pass results in to the constructor of TestJob rather than in here.
  - [x] What's the use case for getting old results?
  - [x] Use await_job_completion here.
  - [x] No need to check status_code.
  - [x] check that they both happen, and in order.

### CODE REVIEW

File: http-service/service.py [hunk 1]
```diff
@@ -253,8 +309,10 @@ def __init__(self, id_, uid, cmd, config, logfile, joblogger):
         self.config = config
         self.logfile = logfile
         self.joblogger = joblogger
-        self.state_data = AtomicTree()
+        self.state_data = AtomicTree(executor=executor)
         self.lock = threading.Condition()
+        self.startTime = datetime.datetime.utcnow()
+        self.endTime = None
```
Comment (wmanley) [inline http-service/service.py]: I believe this should be `end_time`

Comment (drothlis) [inline http-service/service.py]: I'm here

**Comment by wmanley** [comment]
I've just rebased the mongo branch to squash.  Please rebase this branch with:

    git rebase --onto 11760309e8ef22799d047deb45f67b02fc9f9130 \
        95a5a32cddca2bfaadb8bee3e4414ed71f613481

**Comment by drothlis** [comment]
> Please rebase this branch

Done.

**Comment by wmanley** [comment]
I've rebased mongo branch and merged it.  Please:

    git fetch --all --prune
    git rebase --onto wmanley/master 54fc81c

**Comment by drothlis** [comment]
Rebased onto master.

**Comment by wmanley** [comment]
I'd like to see some tests for the start time/end time stuff in service.

File: tests/test_http_service.py [hunk 2]
```diff
@@ -928,3 +928,42 @@ def test_that_existing_results_are_not_reparsed_at_restart(unstarted_runner):
...
+    assert job()['result_counts'] == {
+        "pass": 0, "fail": 0, "error": 0, "total": 2}
+    assert wait_until(lambda: job()['result_counts'] in (
+        {"pass": 1, "fail": 0, "error": 0, "total": 2},
+        {"pass": 1, "fail": 1, "error": 0, "total": 2}))
+    assert wait_until(lambda: job()['status'] != 'running')
+    assert job()['result_counts'] == {
+        "pass": 1, "fail": 1, "error": 0, "total": 2}
+
+    # Check that I can still get old results after running another job
```
Comment (wmanley) [inline tests/test_http_service.py]: What's the use case for getting old results?

File: tests/test_http_service.py [hunk 3]
```diff
@@ -928,3 +928,42 @@ def test_that_existing_results_are_not_reparsed_at_restart(unstarted_runner):
...
         'tests/windows_line_endings.py::test_nothing',
     ]
+
+
+def test_result_counts(runner):
+    result = start_tests(
+        runner, 'aac9006c30beadd38444c3ada11cea93bc35c597',
+        ['tests/example_tests.py::test_sleeping',
+         'tests/example_tests.py::test_failing'])
+    assert result.status_code == 200
```
Comment (wmanley) [inline tests/test_http_service.py]: No need to check `status_code`.  Other tests cover that and this test will fail anyway if we can't get `'job_id'`

File: tests/test_http_service.py [hunk 4]
```diff
@@ -928,3 +928,42 @@ def test_that_existing_results_are_not_reparsed_at_restart(unstarted_runner):
...
+        r = runner.get('/api/_private/jobs/%i' % job_id)
+        assert r.status_code == 200
+        return r.json()
+
+    assert job()['result_counts'] == {
+        "pass": 0, "fail": 0, "error": 0, "total": 2}
+    assert wait_until(lambda: job()['result_counts'] in (
+        {"pass": 1, "fail": 0, "error": 0, "total": 2},
+        {"pass": 1, "fail": 1, "error": 0, "total": 2}))
+    assert wait_until(lambda: job()['status'] != 'running')
```
Comment (wmanley) [inline tests/test_http_service.py]: Use `await_job_completion()` here.  It does a few nice tricks to aid with debugging if something goes wrong.

File: tests/test_http_service.py [hunk 5]
```diff
@@ -928,3 +928,42 @@ def test_that_existing_results_are_not_reparsed_at_restart(unstarted_runner):
...
+    def job():
+        r = runner.get('/api/_private/jobs/%i' % job_id)
+        assert r.status_code == 200
+        return r.json()
+
+    assert job()['result_counts'] == {
+        "pass": 0, "fail": 0, "error": 0, "total": 2}
+    assert wait_until(lambda: job()['result_counts'] in (
+        {"pass": 1, "fail": 0, "error": 0, "total": 2},
+        {"pass": 1, "fail": 1, "error": 0, "total": 2}))
```
Comment (wmanley) [inline tests/test_http_service.py]: I'm not quite sure what the `in` is for here.  You want to check that *either* of these states are reached?  I would have thought you'd want to check that they both happen, and in order.

File: tests/test_http_service.py (hunk 2 above)
Comment (drothlis) [inline tests/test_http_service.py]: None. I've removed that case from the test, though it'll still work.

**Comment by drothlis** [comment]
@wmanley thanks for your review. I've addressed your comments, squashed, and force-pushed.

**Comment by wmanley** [comment]
From test-pack.bundle:

    +def test_sleeping_and_then_failing():
    +    """"This test is a slow test to check that the result_count data is being
    +    updated between test-runs.
    +    """
    +    from time import sleep
    +    sleep(1)
    +    sleep(1)
    +    sleep(1)
    +    assert False

I think we'd get away with sleeping just once for 1 second.

File: tests/test_http_service.py [hunk 6]
```diff
@@ -928,3 +929,99 @@ def test_that_existing_results_are_not_reparsed_at_restart(unstarted_runner):
...
+    # "v0" = before we persisted job information to disk.
+
+    runner = unstarted_runner
+    mkdir_p('log/jobs/000/000/000')
+    runner.start()
+
+    result = runner.get('/api/_private/jobs/%i' % 0)
+    assert result.status_code == 200
+    assert result.json()['id'] == 0
+    assert result.json()['status'] == 'exited'
```
Comment (wmanley) [inline tests/test_http_service.py]: What do you intend 'result_counts' to be in this instance?

**Comment by drothlis** [comment]
I'm not going to write e2e-tests before merging. It's too hard:

* Image matching against the progress bar would need to take into account the animated stripes (while the testcase is running).
* Whether the testcase passes or fails depends on the set-top box that's plugged into the stb-tester ONE appliance under test.

File: tests/test_http_service.py (hunk 6 above)
Comment (drothlis) [inline tests/test_http_service.py]: In the current implementation it simply isn't set, and the front-end displays no progress bar at all (not even the grey progress bar, just an empty grey background).

Comment (drothlis) [inline tests/test_http_service.py]: I pushed a fix so that the web-ui displays a solid grey progress bar in that case.

**Pull request merged**